burstphoto/align.swift

```diff
@@ -40,22 +41,505 @@ let mfl = device.makeDefaultLibrary()!
+    let out_path = (final_dng_conversion ? tmp_dir : out_dir) + out_filename
+    var out_url = URL(fileURLWithPath: out_path)
+       
+    // convert images from uint16 to float16
+    textures = textures.map{texture_uint16_to_float($0)}
+     
+    // set the maximum resolution of the smallest pyramid layer
+    let search_distance_dict = [
+        "Low": 128,
+        "Medium": 64,
+        "High": 32,
+    ]
+    let search_distance_int = search_distance_dict[search_distance]!
+    
+    // use a 32 bit float as final image
+    let final_texture_descriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .r32Float, width: textures[ref_idx].width, height: textures[ref_idx].height, mipmapped: false)
+    final_texture_descriptor.usage = [.shaderRead, .shaderWrite]
+    let final_texture = device.makeTexture(descriptor: final_texture_descriptor)!
+    fill_with_zeros(final_texture)
+    
+    // special mode: simple temporal averaging without alignment and robust merging
+    if noise_reduction == 25.0 {
+        print("Special mode: temporal averaging only...")
+         
+        // correction of hot pixels
+        if mosaic_pattern_width == 2 {
+            correct_hotpixels(textures, black_level)
+            
+            // correction of intensities for images with bracketed exposure
+            // inspired by https://ai.googleblog.com/2021/04/hdr-with-bracketing-on-pixel-phones.html
+            equalize_exposure(textures, black_level, exposure_bias, ref_idx)
+        }
+                  
+        // iterate over all images
+        for comp_idx in 0..<image_urls.count {
+            
+            add_texture(textures[comp_idx], final_texture, image_urls.count)
+            DispatchQueue.main.async { progress.int += Int(80000000/Double(image_urls.count)) }
+        }
+        
+        // apply tone mapping if the reference image is underexposed: by lifting the shadows, more shadow information is available while the tone mapping operator is constructed in such a way that highlights are protected
+        // inspired by https://www-old.cs.utah.edu/docs/techreports/2002/pdf/UUCS-02-001.pdf
+        if (mosaic_pattern_width == 2 && comp_underexposure == "On") {
+            correct_underexposure(final_texture, white_level[ref_idx], black_level, exposure_bias, ref_idx)
+        }
+        
+
+    // sophisticated approach with alignment of tiles and merging of tiles in frequency domain (only 2x2 Bayer pattern)
+    } else if mosaic_pattern_width == 2 {
+        print("Merging in the frequency domain...")
+        
+        // the tile size for merging in frequency domain is set to 8x8 for all tile sizes used for alignment. The smaller tile size leads to a reduction of artifacts at specular highlights at the expense of a slightly reduced suppression of low-frequency noise in the shadows
+        // see https://graphics.stanford.edu/papers/hdrp/hasinoff-hdrplus-sigasia16.pdf for more details
+        let tile_size_merge = Int(8)
```

Comment:
This was originally added (if I remember correctly) in response to some artifacts I noticed in very high contrast areas with blown highlights. I believe that one of your newest commits fixed the issue.

Does this value still need to be hard coded to 8? (If so, then the FFT code is not needed since only the DFT will ever get called)

Reply:
If that the case, could you please add a comment stating as much either on that number or on DFT itself.